File: lab-redes/python/websocket/mural_servidor.py
import asyncio
import websockets

# Parte D - WebSocket (servidor) em Python, com a lib 'websockets'.
# Mural da turma: guarda os conectados e faz broadcast de cada mensagem.
# porta = 8888 + OFFSET(39) = 8927 (ver README)

PORTA = 8927
clientes_conectados = set()


async def tratar_conexao(websocket):
    clientes_conectados.add(websocket)
    print(f"[WebSocket] Novo aluno conectado. Total: {len(clientes_conectados)}")
    await websocket.send("Bem-vindo(a) ao mural de avisos da turma!")
    try:
        async for mensagem in websocket:
            print(f"[WebSocket] Recebido: {mensagem}")
            aviso_formatado = f"Aviso da turma: {mensagem}"
            # manda pra todos os conectados de uma vez
            websockets.broadcast(clientes_conectados, aviso_formatado)
    finally:
        clientes_conectados.remove(websocket)
        print(f"[WebSocket] Aluno desconectado. Total: {len(clientes_conectados)}")


async def main():
    print(f"[WebSocket] Servidor do mural iniciado na porta {PORTA}.")
    async with websockets.serve(tratar_conexao, "0.0.0.0", PORTA):
        await asyncio.Future()  # mantém o servidor rodando pra sempre


asyncio.run(main())

File: lab-redes/java/websocket/src/main/java/MuralServidor.java
import org.java_websocket.WebSocket;
import org.java_websocket.handshake.ClientHandshake;
import org.java_websocket.server.WebSocketServer;

import java.net.InetSocketAddress;

// Parte D - WebSocket (servidor). É o "mural de avisos" da turma: vários alunos
// ficam conectados e, quando um manda uma mensagem, ela é reenviada (broadcast)
// pra todos. O WebSocket começa com um handshake HTTP (Upgrade) e depois mantém
// a conexão TCP aberta pra troca nos dois sentidos, em tempo real.
public class MuralServidor extends WebSocketServer {

    public MuralServidor(int porta) {
        super(new InetSocketAddress(porta));
    }

    @Override
    public void onOpen(WebSocket conexao, ClientHandshake handshake) {
        System.out.println("[WebSocket] Novo aluno conectado: " + conexao.getRemoteSocketAddress());
        conexao.send("Bem-vindo(a) ao mural de avisos da turma!");
    }

    @Override
    public void onMessage(WebSocket conexao, String mensagem) {
        System.out.println("[WebSocket] Recebido: " + mensagem);
        String avisoFormatado = "Aviso da turma: " + mensagem;
        // reenvia pra TODOS os conectados (inclusive quem mandou)
        for (WebSocket cliente : getConnections()) {
            cliente.send(avisoFormatado);
        }
    }

    @Override
    public void onClose(WebSocket conexao, int codigo, String motivo, boolean remoto) {
        System.out.println("[WebSocket] Aluno desconectado: " + conexao.getRemoteSocketAddress());
    }

    @Override
    public void onError(WebSocket conexao, Exception ex) {
        ex.printStackTrace();
    }

    @Override
    public void onStart() {
        System.out.println("[WebSocket] Servidor do mural iniciado.");
    }

    public static void main(String[] args) {
        // porta = 8887 + OFFSET(39) = 8926 (ver README)
        MuralServidor servidor = new MuralServidor(8926);
        servidor.start();
    }
}
